subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultInputProperty.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import com.google.common.base.Objects;
+import org.apache.commons.lang.SerializationUtils;
+import org.gradle.internal.io.ClassLoaderObjectInputStream;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.NotSerializableException;
+import java.io.Serializable;
+
+class DefaultInputProperty implements InputProperty {
+    private transient Object inputProperty;
```

Comment:
I was using it as a cache for 2 reasons:
1. We don't need to deserialize the object if it was already deserialized before.
2. In order to deserialize the input property, we need the right class loader (which was the original issue). In the implemented solution, we know that at least one of the two `DefaultInputProperty` will have the instance of a deserialized input property (as it's the current build session). We simply use the class loader of that object to deserialize the other one.

Reason 1 is probably negligible. Could you elaborate on how to solve reason 2 without causing `ClassNotFoundException`?